tests/sentry/ingest/test_transaction_clusterer.py

```diff
@@ -27,3 +39,69 @@ def test_single_leaf():
     ]
     clusterer.add_input(transaction_names)
     assert clusterer.get_rules() == ["/a/*/**"]
+
+
+@mock.patch("sentry.ingest.transaction_clusterer.datasource.redis.MAX_SET_SIZE", 5)
+def test_collection():
+    project1 = Project(id=101, name="p1", organization_id=1)
+    project2 = Project(id=102, name="project2", organization_id=1)
+
+    for project in (project1, project2):
+        for i in range(len(project.name)):
+            _store_transaction_name(project, f"tx-{project.name}-{i}")
+            _store_transaction_name(project, f"tx-{project.name}-{i}")
+
+    set_entries1 = _get_transaction_names(project1)
+    assert set(set_entries1) == {"tx-p1-0", "tx-p1-1"}
+
+    set_entries2 = _get_transaction_names(project2)
+    assert len(set_entries2) == 5
+    # We don't know which entries made it into the final set:
+    for name in set_entries2:
+        assert name.startswith("tx-project2-")
+
+    project3 = Project(id=103, name="project3", organization_id=1)
+    assert set() == _get_transaction_names(project3)
+
+
+@mock.patch("sentry.ingest.transaction_clusterer.datasource.redis.MAX_SET_SIZE", 100)
+def test_distribution():
+    """Make sure that the redis set prefers newer entries"""
```

Comment:
The lua script uses `SPOP` to remove items when there's an overflow. From the [docs](https://redis.io/commands/spop/):

> Removes and returns one or more random members from the set value store at key.

We can't ensure redis prefers new entries over old ones. Or am I missing something?

Reply:
> an item's chance of survival decreases with every insert, such that overall, the set will be biased towards newer entries

You're assuming that the "random" in the redis docs actually means random, and it doesn't rely on the item's hash. For example, it could always remove the first/last item in the set, where items are sorted by their hash.